### DISCUSSION TITLE
preserve topological build order (via ^) but have exceptions

### DISCUSSION THREAD

**Discussion by is-it-ayush** [vercel/turborepo#6203] [Help] [answered]
Hi, In `turbo.json`, you can use `dependsOn` attribute to run commands before a command is ran. For example,
```
{
  "build": {
    dependsOn: ["^build", "test"],
  }
}
``` 
will run build each package's `dependencies` & `devDependencies` before building the package itself due to `^build`. Is there a way to do the same but ignore certain dependencies. I have a dependency that does some code-gen and doesn't need to built again. I want to ignore it here while still preserving this topological build order. (I have more dependencies that need building). I looked at the documentation for `dependsOn` but I couldn't find anything. : )

**Comment by anthonyshew** [comment]
Is hitting cache for the task insufficient?

**Reply by is-it-ayush** [reply]
Hi Anthony!! Thank you for the response. As far as I remember the problem was I needed to stop a dependency from being built. This dependency was supposed to be built by another CI workflow running a very specific operating system. The dependency could only compile on it otherwise compilation failed. 

> Is hitting cache for the task insufficient?

Hitting the cache wasn't possible here (due to a different workflow). Turbo did it's job & attempted to build this specific dependency but it failed the entire CI. I was hoping for an `ignore: []` array similar to `dependsOn: []` whereby I could specify dependency `x` to be ignored but the rest of the topological order to be persevered. I read the docs but I was struck with the realisation that it's not possible to tell turbo "Hey! Please don't build this dependency. It won't compile in your CI env".

**Reply by anthonyshew** [reply]
I see. The answer here will be no, we don't have a way to do precisely what you're describing. However, I think you might have options:
- Create another pipeline, leaving out the dependencies you don't want
- If the dependency you want to exempt is a specific workspace, you can use `--filter`: https://turbo.build/repo/docs/reference/command-line-reference/run#--filter

**Reply by is-it-ayush** [reply]
Thank you @anthonyshew! I haven't tried the negated filter yet but quick question, 
- Say we have two packages, `@acme/specific_os_compile_only` and `@acme/web` where the latter depends on the former (topological build order; according to turbo). If we were to run `turbo run build --filter=!@acme/specific_os_compile_only`, Would the topological order still force `@acme/specific_os_compile_only` to be built before `@acme/web` is built even with the negated filter flag?

Tiny request: Could we have `ignoreDeps: []` as a feature please!! I understand this is very edge case I ran into but it'd be awesome if we could tell "Preserve my build order for package 'x' but ignore that dependency 'y' on package 'x'".

**Reply by anthonyshew** [reply]
Quick answer to your filtering question, you could do: `turbo run build --filter=web --filter=!@acme/specific_os_compile_only` and I believe that would take care of what you're describing.

Longer answer...

Thinking about `turbo` from core principles, it seems to me that something like `ignoreDeps` would be problematic:
- Fundamentally, what an exception like that is describing is an entirely separate pipeline with different requirements. The topology of your repo is the key to understanding what to build, when to build, what to cache, when to cache...Almost everything. To throw out a dependency for that pipeline in `turbo.json` seems like a workaround to a monorepo whose dependency tree might need a second thought. (I say this in the general sense - might not be the case for your repo).
- Building on that idea, I'm trying to think of `ignoreDeps` (or its equivalent) as a generalized solution that goes beyond more than this specific use case. The more I think about the ways that Turborepo users could use this API, it sounds like a _massive_ footgun. One could very quickly build pipelines that, among other problems, are extremely confusing, can deploy cached code that shouldn't have been, etc.
- You mentioned that the task needs a "very specific operating system". To me, _this is the real key_. The true solution here seems to be the `turbo` would be able to hold an understanding of OS in its model. (We don't have work planned for this today - but it does seem that it would pave the way to solving this problem).
- It seems that this could also be solved by making a quick script in the package's `package.json` that says "if (TARGET_OS) build, else exit 0". This would make it so that that task passes, at least, but you may find you need to apply `cache: false` to this specific task as a result of this strategy. I don't know the deeper specifics of your repo so can't speak a perfect solution.

Ultimately, I want to make sure the team thinks through an API like this heavily before making it a generalized solution to a non-general problem.

**Reply by is-it-ayush** [reply, marked as answer]
Hi @anthonyshew! Thank you for the detailed answer. 

> The more I think about the ways that Turborepo users could use this API, it sounds like a massive footgun.

I agree! `ignoreDeps` might introduce side-effects into turbo's pipeline which could negatively effect an already working and clearly defined pipeline.

> It seems that this could also be solved by making a quick script in the package's package.json that says "if (TARGET_OS) build, else exit 0". This would make it so that that task passes, at least, but you may find you need to apply cache: false to this specific task as a result of this strategy. I don't know the deeper specifics of your repo so can't speak a perfect solution.

Thank you! The shell script hack works. [Here's the change I made](https://github.com/is-it-ayush/turbo-6203-repro/pull/1). I believe this to be a workaround rather a solution since turbo would still attempt to run build script on all packages without this shell script hack. 

To better represent this issue, I hacked together a quick repro that is a close representation of the problem. This will help you better understand this edge case Anthony. [Here is it](https://github.com/is-it-ayush/turbo-6203-repro)!
